plugins/preserve-tombs.cpp

```diff
@@ -0,0 +1,269 @@
+        is_enabled = enable;
+        DEBUG(config,out).print("%s from the API; persisting\n",
+                                is_enabled ? "enabled" : "disabled");
+        set_config_bool(config, CONFIG_IS_ENABLED, is_enabled);
+        EventManager::registerListener(EventManager::EventType::UNIT_DEATH, assign_tomb_handler, plugin_self);
+        if (enable)
+            update_tomb_assignments(out);
+    } else {
+        EventManager::unregisterAll(plugin_self);
+        DEBUG(config,out).print("%s from the API, but already %s; no action\n",
+                                is_enabled ? "enabled" : "disabled",
+                                is_enabled ? "enabled" : "disabled");
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown (color_ostream &out) {
+    DEBUG(config,out).print("shutting down %s\n", plugin_name);
+
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_load_data (color_ostream &out) {
+    cycle_timestamp = 0;
+    config = World::GetPersistentData(CONFIG_KEY);
+
+    if (!config.isValid()) {
+        DEBUG(config,out).print("no config found in this save; initializing\n");
+        config = World::AddPersistentData(CONFIG_KEY);
+        set_config_bool(config, CONFIG_IS_ENABLED, is_enabled);
+        set_config_val(config, CONFIG_CYCLES, 25);
+    }
+
+    is_enabled = get_config_bool(config, CONFIG_IS_ENABLED);
+    cycle_freq = get_config_val(config, CONFIG_CYCLES);
+    DEBUG(config,out).print("loading persisted enabled state: %s\n",
+                            is_enabled ? "true" : "false");
+
+    if (is_enabled) update_tomb_assignments(out);
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    if (event == DFHack::SC_WORLD_UNLOADED) {
+        tomb_assignments.clear();
+        if (is_enabled) {
+            DEBUG(config,out).print("world unloaded; disabling %s\n",
+                                    plugin_name);
+            is_enabled = false;
+        }
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out) {
+    if (is_enabled && world->frame_counter - cycle_timestamp >= cycle_freq)
+        update_tomb_assignments(out);
+    return CR_OK;
+}
+// </BOILERPLATE>
+
+
+
+
+// On unit death - check if we assigned them a tomb
+//
+//
+void onUnitDeath(color_ostream& out, void* ptr) {
+    // input is void* that contains the unit id
+    int32_t unit_id = reinterpret_cast<std::intptr_t>(ptr);
+
+    // check if unit was assigned a tomb in life
+    auto it = tomb_assignments.find(unit_id);
+    if (it == tomb_assignments.end()) return;
+
+    // assign that unit to their previously assigned tomb in life
+    int32_t building_id = it->second;
+    if (!assign_to_tomb(unit_id, building_id)) return;
+
+    // success, print status update and remove assignment from our memo-list
+    out.print("Unit %d died - assigning them to tomb %d\n", unit_id, building_id);
```

Comment:
this should be an INFO statement instead of a raw print (so players have control over whether it is output)